src/librustc/infer/error_reporting/nice_region_error/placeholder_error.rs

```diff
@@ -218,88 +246,102 @@ impl NiceRegionError<'me, 'gcx, 'tcx> {
+                expected_has_vid = Some(counter);
+                counter += 1;
+            }
         });
 
-        self.tcx.for_each_free_region(&actual_trait_ref, |r| {
-            if Some(r) == vid && has_vid.is_none() {
-                has_vid = Some(counter);
+        self.tcx().for_each_free_region(&actual_trait_ref, |r| {
+            if Some(r) == vid && actual_has_vid.is_none() {
+                actual_has_vid = Some(counter);
                 counter += 1;
             }
         });
 
         let self_ty_has_vid = self
-            .tcx
+            .tcx()
             .any_free_region_meets(&actual_trait_ref.self_ty(), |r| Some(r) == vid);
 
         RegionHighlightMode::maybe_highlighting_region(sub_placeholder, has_sub, || {
             RegionHighlightMode::maybe_highlighting_region(sup_placeholder, has_sup, || {
                 match (has_sub, has_sup) {
                     (Some(n1), Some(n2)) => {
                         err.note(&format!(
-                            "`{}` must implement `{}` \
-                             for any two lifetimes `'{}` and `'{}`",
-                            expected_trait_ref.self_ty(),
+                            "`{}` would have to be implemented for the type `{}`, \
+                            for any two lifetimes `'{}` and `'{}`",
                             expected_trait_ref,
+                            expected_trait_ref.self_ty(),
                             std::cmp::min(n1, n2),
                             std::cmp::max(n1, n2),
                         ));
                     }
                     (Some(n), _) | (_, Some(n)) => {
                         err.note(&format!(
-                            "`{}` must implement `{}` \
-                             for any lifetime `'{}`",
-                            expected_trait_ref.self_ty(),
+                            "`{}` would have to be implemented for the type `{}`, \
+                            for any lifetime `'{}`",
                             expected_trait_ref,
+                            expected_trait_ref.self_ty(),
                             n,
                         ));
                     }
                     (None, None) => {
                         err.note(&format!(
-                            "`{}` must implement `{}`",
-                            expected_trait_ref.self_ty(),
+                            "`{}` would have to be implemented for the type `{}`",
                             expected_trait_ref,
+                            expected_trait_ref.self_ty(),
                         ));
                     }
                 }
             })
         });
 
-        RegionHighlightMode::maybe_highlighting_region(vid, has_vid, || match has_vid {
-            Some(n) => {
-                if self_ty_has_vid {
+        RegionHighlightMode::maybe_highlighting_region(
+            vid,
+            actual_has_vid.or(expected_has_vid),
+            || match actual_has_vid {
```

Comment:
I don't think it would add a lot of value to refactor out personally, though I'm not opposed or anything.

Reply:
I'm trying to say *why* .. I guess because this just feels like a big block of procedural code. Maybe some comments explaining what it is doing at each point. 

(e.g., "// explain about the impl we found")

otoh maybe having a fn named `explain_actual_impl_that_was_found` is nicer. either way.

Reply:
I think it would be nicer because I can then just read `explain_actual_impl_that_was_found(...)` and not have to bother thinking about how it does what it does which reduces overall cognitive load... (this is a general concern I have with most code in this repo in that it makes things unreadable).

Reply:
sure, I'll try extracting a helper function.